### ISSUE TITLE
Investigate Freedesktop SDK

### ISSUE THREAD

**Issue by probonopd** [AppImage/AppImageKit#992]
Investigate Freedesktop SDK to build what goes into an AppImage.

Freedesktop SDK claims to be distribution agnostic, desktop environment agnostic, and package format agnostic:

![agnostic](https://user-images.githubusercontent.com/2480569/64919533-7dae7a00-d79b-11e9-9e20-744faa6cc878.png)

* So how do we compile something using Freedesktop SDK without Flatpak?

**Comment by probonopd** [comment]
Note to self:

```
sudo add-apt-repository -y ppa:alexlarsson/flatpak
sudo apt update

sudo apt-get -y install flatpak
flatpak remote-add --if-not-exists flathub https://flathub.org/repo/flathub.flatpakrepo
flatpak install flathub org.freedesktop.Sdk//19.08
flatpak install flathub org.freedesktop.Platform//19.08
mksquashfs /var/lib/flatpak/runtime/org.freedesktop.Platform/x86_64/19.08/active/files/ org.freedesktop.Platform_19.08_x86_64.sqs
```

Further investigation blocked because [I can't get Flatpak to work](https://github.com/flatpak/flatpak/issues/2200#issuecomment-531551643), as has been my problem since its inception.

**Comment by probonopd** [comment]
Reference:
https://gitlab.com/freedesktop-sdk/freedesktop-sdk/issues/881

**Comment by Wetrain** [comment]
-     So how do we compile something using Freedesktop SDK without Flatpak? Let's start with GNU Hello https://www.gnu.org/software/hello/

  - You can use BuildStream to do so: https://docs.buildstream.build/1.4.1/tutorial/first-project.html

-     Which version of Freedesktop SDK do we need to use so that binaries compiled using it will run with the Platform "runtime" that comes by default with all still-supported distributions? (I mean, not a real Freedesktop runtime but the OS itself.)

  - I'm not to sure on this requirement, but we maintain backwards/forwards compatibility for 2 years      with each release 

-     In case we would allow a Freedesktop Platform (runtime) to be required to run an AppImage, how large would the recent Freedesktop Platform (runtime) be (as a squashfs image)? --> mksquashfs /var/lib/flatpak/runtime/org.freedesktop.Platform/x86_64/19.08/active/files/ org.freedesktop.Platform_19.08_x86_64 209 MB, Sdk 520 MB

   - seems you have already done the calculations here, but these seems correct?

-     Why is Snappy not using Freedesktop Platforms but their own Base Snaps?
  -  because they use their own Snap runtime out of choice, however we have worked with them to get freedesktop-sdk accepted as a community runtime https://forum.snapcraft.io/t/base-runtime-freedesktop-sdk-runtime-19-08/11153/21

-     What would the advantage be to use Freedesktop Platforms rather than Base Snaps?
  -  We are a distribution agnostic runtime and are built entirely from source, with 19.08 being self bootstrapped.

-     How often do new versions of the Freedesktop Platform appear? (Hopefully only once after every few years!)

  - https://gitlab.com/freedesktop-sdk/freedesktop-sdk/wikis/release

-     Is there a backward binary compatibility guarantee, e.g., saying that binaries compiled with Freedesktop SDK version i will run on Freedesktop Platform i, i+1, ... i+n? What is n?

  - see above ^

-     What is org.freedesktop.Platform.Compat? it is undocumented on http://docs.flatpak.org/en/latest/available-runtimes.html

  - It's a 32 bit extension, for example which is used by Steam to support old games

I hope these answer some of your questions, i would encourage you to ask some questions in the channel on freenode (#freedesktop-sdk) also as some developers are more active there than on Gitlab.

**Comment by probonopd** [comment]
Thank you very much for your answers @Wetrain. Highly appreciated.

Especially the page you linked to,

> The aim is to continuously update the runtime while keeping the ABI backward and forward compatible for as long as reasonably possible.
> 
> For the base runtime we expect multiple years of ABI stability to be feasible (parallel installation of multiple versions may occasionally be necessary for a few libraries). When an ABI break is required (e.g., C++ ABI break), the plan is to keep the previous branch maintained (at least for security updates) for two years.

is highly relevant to us. Will have to think about the implications a bit and discuss.
